components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+            self.remove(range[0], range[1]);
+        }
+    }
+
+    /// Retain the elements in the UnicodeSetBuilder that are in the parameter
+    ///
+    /// Performs this in an `O(R)` amortized run-time, where `R` is the number of ranges
+    fn retain(&mut self, ret: Vec<u32>) {
+        let mut chunk = vec![0];
+        chunk.extend(ret);
+        chunk.push((char::MAX as u32) + 1);
+        chunk.chunks(2).for_each(|c| self.remove(c[0], c[1]));
+    }
+
+    /// Retain the specified character in the UnicodeSetBuilder if it exists
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSetBuilder;
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// builder.add_range(&('A'..'Z'));
+    /// builder.retain_char('A');
+    /// let set = builder.build();
+    /// let mut check = set.iter();
+    /// assert_eq!(check.next(), Some('A'));
+    /// assert_eq!(check.next(), None);
+    /// ```
+    pub fn retain_char(&mut self, c: char) {
+        let code_point = c as u32;
+        let to_retain = vec![code_point, code_point + 1];
+        self.retain(to_retain);
+    }
+
+    /// Retain the range of characters located within the UnicodeSetBuilder
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSetBuilder;
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// builder.add_range(&('A'..'Z'));
+    /// builder.retain_range(&('A'..='B'));
+    /// let set = builder.build();
+    /// let mut check = set.iter();
+    /// assert_eq!(check.next(), Some('A'));
+    /// assert_eq!(check.next(), Some('B'));
+    /// assert_eq!(check.next(), None);
+    /// ```
+    pub fn retain_range(&mut self, range: &impl RangeBounds<char>) {
+        let (start, end) = deconstruct_range(range);
+        let to_retain = vec![start, end];
+        self.retain(to_retain);
+    }
+
+    /// Retain the elements in the specified set within the UnicodeSetBuilder
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::{UnicodeSetBuilder, UnicodeSet};
+    /// let mut builder = UnicodeSetBuilder::new();
+    /// let set = UnicodeSet::from_inversion_list(vec![65, 70]).unwrap();
+    /// builder.add_range(&('A'..'Z'));
+    /// builder.retain_set(&set); // retains 'A'..'F'
+    /// let check = builder.build();
+    /// assert!(check.contains('A'));
+    /// assert!(!check.contains('G'));
+    /// ```
+    pub fn retain_set(&mut self, set: &UnicodeSet) {
+        self.retain(set.ranges().collect::<Vec<&[u32]>>().concat());
+    }
+
+    /// Computes the complement of the argument, adding any elements that do not yet exist in the builder,
+    /// and removing any elements that already exist in the builder. See public functions for examples.
+    ///
+    /// Performs in `O(B + S)`, where `B` is the number of endpoints in the Builder, and `S` is the number
+    /// of endpoints in the argument.
+    fn complement_list(&mut self, set: Vec<u32>) {
```

Comment:
Take a `&[u32]` slice instead of a Vec by value, which lets you be more efficient at the call sites.

Reply:
agreed and changed